### PR TITLE
Monorepo/app store MS Teams Integration

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR adds the option to create a MS Teams link as a event location. Note that this only works with enterprise accounts (ie. work or school)

Addresses #1462

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

Setup
- Ensure that you created MS Graph credentials in your .env file
- In Azure -> App Registration -> "Your app" -> API Permissions, add the `
OnlineMeetings.ReadWrite` permission 
- Enroll in the MS Developer program to gain access to a test enterprise account
    - Link: https://developer.microsoft.com/en-us/microsoft-365/dev-program
    - I can also provide the account that I used

Testing
 - Create a new user 
 - Once logged in go to the url `http://localhost:3000/apps/installed`
 - Under "Conferencing" connect to MS Teams
 - On the MS sign in page enter the test enterprise account details
 - Once connected create a new event type and set the location to MS Teams
 - With the event type, create a new booking. You should see that the meeting URL is created

### CODE REVIEW

File: apps/web/components/booking/pages/BookingPage.tsx [hunk 1]
```diff
@@ -144,6 +144,7 @@ const BookingPage = (props: BookingPageProps) => {
     [LocationType.Daily]: "Daily.co Video",
     [LocationType.Huddle01]: "Huddle01 Video",
     [LocationType.Tandem]: "Tandem Video",
+    [LocationType.Teams]: "MS Teams",
```
Comment (joeauyeung) [inline apps/web/components/booking/pages/BookingPage.tsx:147]: Add MS Teams to location options

File: apps/web/lib/events/EventManager.ts [hunk 2]
```diff
@@ -53,8 +53,14 @@ export const isTandem = (location: string): boolean => {
...
 
+export const isTeams = (location: string): boolean => {
+  return location === "integrations:office365_video";
+};
+
 export const isDedicatedIntegration = (location: string): boolean => {
-  return isZoom(location) || isDaily(location) || isHuddle01(location) || isTandem(location);
+  return (
+    isZoom(location) || isDaily(location) || isHuddle01(location) || isTandem(location) || isTeams(location)
+  );
```
Comment (joeauyeung) [inline apps/web/lib/events/EventManager.ts:63]: Add checks for MS Teams integrations

File: apps/web/lib/events/EventManager.ts [hunk 3]
```diff
@@ -112,11 +119,11 @@ export default class EventManager {
...
-    const hasDailyIntegration = process.env.DAILY_API_KEY;
-    if (hasDailyIntegration) {
-      this.videoCredentials.push(FAKE_DAILY_CREDENTIAL);
-    }
-    this.videoCredentials.push(FAKE_HUDDLE_CREDENTIAL);
+    // const hasDailyIntegration = process.env.DAILY_API_KEY;
+    // if (hasDailyIntegration) {
+    //   this.videoCredentials.push(FAKE_DAILY_CREDENTIAL);
+    // }
+    // this.videoCredentials.push(FAKE_HUDDLE_CREDENTIAL);
```
Comment (joeauyeung) [inline apps/web/lib/events/EventManager.ts:126]: @zomars I know you were working on migrating the Daily integration so not sure if these are needed

File: apps/web/pages/api/integrations/office365video/callback.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    data: {
+      type: "office365_video",
+      key: responseBody,
+      userId: session.user.id,
+    },
+  });
+
+  const state = decodeOAuthState(req);
+  return res.redirect(state?.returnTo ?? "/integrations");
+}
```
Comment (joeauyeung) [inline apps/web/pages/api/integrations/office365video/callback.ts:73]: Under `pages/api` redirect back to our app and grab tokens

File: apps/web/pages/event-types/[type].tsx [hunk 5]
```diff
@@ -602,6 +604,78 @@ const EventTypePage = (props: inferSSRProps<typeof getServerSideProps>) => {
...
+                          d="M95.01,466.5h950.312c52.473,0,95.01,42.538,95.01,95.01v950.312c0,52.473-42.538,95.01-95.01,95.01	H95.01c-52.473,0-95.01-42.538-95.01-95.01V561.51C0,509.038,42.538,466.5,95.01,466.5z"
+                        />
+                        <path
+                          fill="#FFF"
+                          d="M820.211,828.193H630.241v517.297H509.211V828.193H320.123V727.844h500.088V828.193z"
+                        />
+                      </svg>
+                      <span className="ml-2 text-sm">MS Teams</span>
+                    </div>
+                  )}
```
Comment (joeauyeung) [inline apps/web/pages/event-types/[type].tsx:678]: Added Teams icon to location list

File: packages/app-store/office365video/api/add.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+      scope: scopes.join(" "),
+      client_id: process.env.MS_GRAPH_CLIENT_ID,
+      redirect_uri: BASE_URL + "/api/integrations/office365teams/callback",
+      // state,
+    };
+    const query = stringify(params);
+    const url = `https://login.microsoftonline.com/common/oauth2/v2.0/authorize?${query}`;
+    res.status(200).json({ url });
+  }
+}
```
Comment (joeauyeung) [inline packages/app-store/office365video/api/add.ts:22]: I noticed that the Zoom package still made calls to the API. I added the API endpoints under packages in preparation for the migration to the app store.

File: packages/app-store/office365video/utils.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+export function decodeOAuthState(req: NextApiRequest) {
+  if (typeof req.query.state !== "string") {
+    return undefined;
+  }
+  const state: IntegrationOAuthCallbackState = JSON.parse(req.query.state);
+
+  return state;
+}
```
Comment (joeauyeung) [inline packages/app-store/office365video/utils.ts:21]: Needed some functions from util so copied it under the package.

File: packages/app-store/office365video/.env.example [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# Place these on `apps/web/.env`
+# @see https://github.com/calcom/cal.com/#obtaining-zoom-client-id-and-secret
```
Comment (zomars) [inline packages/app-store/office365video/.env.example:2]: Let's update this URL

File: apps/web/pages/api/cancel.ts [hunk 9]
```diff
@@ -154,9 +154,9 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
     },
   });
 
-  if (bookingToDelete.location === "integrations:daily") {
-    bookingToDelete.user.credentials.push(FAKE_DAILY_CREDENTIAL);
-  }
+  // if (bookingToDelete.location === "integrations:daily") {
+  //   bookingToDelete.user.credentials.push(FAKE_DAILY_CREDENTIAL);
+  // }
```
Comment (zomars) [inline apps/web/pages/api/cancel.ts:159]: We still need this for now. Let's handle this in a follow up.

File: apps/web/pages/api/integrations/office365video/add.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
+import type { NextApiRequest, NextApiResponse } from "next";
```
Comment (zomars) [inline apps/web/pages/api/integrations/office365video/add.ts:1]: We can remove this file and just keep the one in `packages/app-store/office365video/api/add.ts`

**Review by pumfleet — COMMENTED** [review]
Code looks good, will test functionality tomorrow (it's late now)

**Review by leog — APPROVED** [review]
Well done, I couldn’t find anything to comment on. Also a good idea to comment your own changes to guide the review a bit better, I will apply it myself 😄

**Comment by zomars** [comment]
> Well done, I couldn’t find anything to comment on. Also a good idea to comment your own changes to guide the review a bit better, I will apply it myself 😄

Agreed @leog, you can see the "How to Perform a Self-Review" in our Twist Engineering channel.

**Review by zomars — APPROVED** [review]

**Pull request merged**